modules/web3/events.ts

```diff
@@ -0,0 +1,131 @@
+import { Event } from '@ethersproject/contracts';
+import { Interface } from '@ethersproject/abi';
+import { chunk } from 'lodash';
+
+export const getEvents = async (
+    fromBlock: number,
+    toBlock: number,
+    addresses: string[],
+    topics: string[],
+    rpcUrl: string,
+    rpcMaxBlockRange: number,
+    abi?: any,
+): Promise<Event[]> => {
+    let iEvents: Interface;
+    if (abi && abi.length > 0) {
+        iEvents = new Interface(abi);
+        // check if topics are event names
+        const alreadyEncoded = topics.every((topic) => topic.startsWith('0x'));
+        if (!alreadyEncoded) topics = topics.map((topic) => iEvents.getEventTopic(topic));
+    }
+
+    const range = toBlock - fromBlock;
+    const numBatches = Math.ceil(range / rpcMaxBlockRange);
+
+    const promises: Promise<Event[]>[] = [];
+
+    for (let i = 0; i < numBatches; i++) {
+        const from = fromBlock + (i > 0 ? 1 : 0) + i * rpcMaxBlockRange;
+        const to = Math.min(fromBlock + (i + 1) * rpcMaxBlockRange, toBlock);
+
+        const addressChunks = chunk(addresses, 500);
+
+        for (const addressChunk of addressChunks) {
+            const promise = fetchLogs(from, to, addressChunk, topics, rpcUrl).catch((e: any) => {
+                // Ankr RPC returns error if block range is too wide
+                if (e.includes('block range is too wide')) {
+                    return getEvents(from, to, addressChunk, topics, rpcUrl, rpcMaxBlockRange / 2);
+                }
+
+                // Infura returns 'more than 10000 results' error if block range is too wide
+                if (e.includes('query returned more than 10000 results')) {
+                    const range = e
+                        .match(/\[([0-9a-fA-F, x]+)\]/)
+                        .pop()
+                        .split(', ')
+                        .map((hex: string) => parseInt(hex, 16));
+
+                    return getEvents(from, to, addressChunk, topics, rpcUrl, range[1] - range[0]);
```

Comment:
Could you describe the actual string you are parsing here in the comments? Hard to say whether the string manipulation is correct